New Feature - Build Diff Viewer for asset comparison

Masusder opened the pull request:
I have implemented a Diff Viewer for side-by-side comparison and visualization of assets from two game builds.

To add it I had to refactor core of the app unfortunately, as obviously FModel was designed to load and work with only a single game build at a time. Although, I think many would love to have this feature.

Below is quick overview of everything I added:

- Settings, ability to load second game build (can be null). I made it an expandable button, so it won't bother anyone who isn't interested in this option. It's saved to app settings just as usual game build would be, mappings can be only set after first save.

![image](https://github.com/user-attachments/assets/381780ca-9cad-4ce0-a604-767954fe0911)

- Diff viewer option. It's disabled by default if second game build isn't configured.

![image](https://github.com/user-attachments/assets/e916ca99-b9ec-4dfd-bcf7-e8350331f322) ![image](https://github.com/user-attachments/assets/65b1e1f9-4cfe-4aac-94f3-0ab61e0106cc)

- Diffs are computed via DiffPlex library, for large assets (150k+ lines) I've added chunking to handle the quadratic complexity problem (O(N²)), more chunks are loaded as you scroll to the bottom. Unfortunately this approach loses full diff aligment, but it's better than not having support for large assets at all.

![image](https://github.com/user-attachments/assets/2dfdd48a-9b51-4011-b0a7-9bc90b1e313e)
![image](https://github.com/user-attachments/assets/e4a29167-7ec9-4609-8c3a-fee7da2566c8)

- Viewer comes with expandable navigation bar on the right side which you can use to quickly find content that was changed.

![image](https://github.com/user-attachments/assets/70caff4d-c7dd-4ac3-aaf9-f57cc0f93fe0)

- I've reused searchbar from AvalonEditor.

![image](https://github.com/user-attachments/assets/98eca458-1434-470f-a686-9e36f47eb4b1)

- In case compared assets are identical, appropiate message will be shown. Images are compared by visual hash approach instead, because I don't want to show users images that visually look identical.

![image](https://github.com/user-attachments/assets/80bc8672-5319-4aeb-bac9-5742ae5fead6)

- If given asset path doesn't exist anymore, but asset with the same name exists in different directory, user will be given option to select one of these files manually. That's because file might have been simply moved and still exists under different directory.

![image](https://github.com/user-attachments/assets/9eb427e2-1cdc-4067-aeb3-24f907605c83)

- Backups are now divided into two categories - Lite/Heavy. Lite backup works just like before, while Heavy backup copies currently selected game directory into folder of user choice. User will be notified how much space it's going to take and in case there's not enough warning will be displayed instead. Backup creation buttons are now blocked until backup is finished.

![image](https://github.com/user-attachments/assets/bc0c4af2-582b-4a8e-b5df-e4e39ab717bd) ![image](https://github.com/user-attachments/assets/24ad5828-6e96-4bda-be1b-10186eabcad0)
![image](https://github.com/user-attachments/assets/651369a2-92e2-49b7-98b1-d79f4eb3aee1)

Other info:
I've split CUE4ParseViewModel into multiple files, simply because it was huge and this way it's easier to maintain in my opinion, especially that now it handles two Providers.
~~I haven't tested any game that uses dynamic AES keys, so no idea if AES settings will work for these.~~

There's a chance I might have broken something along the way and haven't noticed (mainly talking about Provider logic), so it would be great if you let me know.

I really hope this PR could be included officially. I shared it to a couple of people for testing and they all found this feature very useful.

Masusder committed:
Diff viewer prototype

Big mess right now, but working

Masusder committed:
Improved diff viewer, diffs are now aligned

Still need to add searchbar

Masusder committed:
Glorified diff settings UI + it's user-friendly

Proto image diff viewer
View Diff option won't be enabled if compared build isn't set
Improved diff tabs behavior

(I still need to add AES settings for compared build)

Masusder committed:
Searchbar, diff loaded virtually by chunks (for large diffs), async initialization, fixed gap brush artifacts

Masusder committed:
AES settings, UE version fix, no image fallback, identical data handler

Masusder committed:
Multiple diffs selection, fixed tab content presenter

Masusder committed:
Extract once

Masusder committed:
Read settings

Masusder committed:
Lite/Heavy backup

Masusder committed:
Improved backup creation UX and validation

Masusder committed:
Improved diff game settings handling and UI feedback

Masusder committed:
Let's just copy whole folder instead + cancellation

Masusder committed:
Better modified lines highlighting

Masusder committed:
Base for navbar (unfinished)

Masusder committed:
Diff navigation bar!

Masusder committed:
Add DiffPlex to third-party notices

Masusder committed:
Scroll faster on navbar and adjust when size changes

Masusder committed:
Sync

Masusder committed:
Merge remote-tracking branch 'upstream/dev' into feature/build-diff-viewer

Masusder committed:
Update navbar for chunks, fixed feedback loop

Masusder committed:
Only navbar height change matters

Masusder committed:
Fixed gap brush not scrolling horizontally

Masusder committed:
Centered scroll indicator

Masusder committed:
Added visual hash comparison for image diffs

Moved search bar logic from AvalonEditor to AvalonSearchbar (which I previously isolated for diff viewer)

Masusder committed:
Don't update navbar when it's not needed

Masusder committed:
Merge branch '4sval:dev' into feature/build-diff-viewer

Masusder committed:
Merge remote-tracking branch 'upstream/dev' into feature/build-diff-viewer

Masusder committed:
Merge remote-tracking branch 'upstream/dev' into feature/build-diff-viewer

Masusder committed:
Merge from commit aa75a8b

AllyJaxx commented:
<img width="1055" height="217" alt="image" src="https://github.com/user-attachments/assets/caa17d4f-e4a9-4295-9c21-f21082ce5d17" />
<img width="1528" height="970" alt="image" src="https://github.com/user-attachments/assets/b57efaac-4b1d-4388-ac2e-e8d3fb70efb0" />

As you suggested, games with AES keys (in this case an old version of Fortnite) do not work as expected here, with no option to add a key either as far as I can tell.

4sval commented:
the dedication to make this work on top of the current outdated architecture is crazy
so is the time the review will take

Masusder committed:
Dynamic AES keys

Masusder commented:
@AllyJaxx Dynamic AES keys configuration should work now, thank you for the samples

Masusder committed:
Merge remote-tracking branch 'upstream/dev' into feature/build-diff-viewer

Masusder committed:
Merge remote-tracking branch 'upstream/dev' into feature/build-diff-viewer

Masusder committed:
Extracted common logic

Masusder committed:
Merge upstream

Masusder committed:
Added compared game selector

Masusder committed:
Fixed versions filtering + improved navbar aligment

Reduced minimum height of markers and scroll indicator to more accurately represent proportional scroll position

Masusder committed:
Merge upstream

Masusder committed:
Improved chunk loading and game selector

Masusder committed:
Ability to select chunks

Masusder commented:
I made some further improvements

Compared game build can now also be selected from directory selector (feature requested by @AllyJaxx):

<img width="484" height="452" alt="NzhEpP5" src="https://github.com/user-attachments/assets/8ccefea1-8e6e-4c8a-ab79-d137a62ee29f" />

Instead of forcing loading of diff chunks when scrolled to the bottom, users can now choose whether to load more chunks and how many to load at a time:

<img width="457" height="125" alt="image" src="https://github.com/user-attachments/assets/f0dc4abb-34c1-4b71-87c2-7becb43a0e71" />

Masusder committed:
Merge remote-tracking branch 'upstream/dev' into feature/build-diff-viewer

Masusder committed:
Merge remote-tracking branch 'upstream/dev' into feature/build-diff-viewer

Masusder committed:
Fixed searchbar not focusing

Masusder committed:
Merge remote-tracking branch 'upstream/dev' into feature/build-diff-viewer

Masusder committed:
Merge remote-tracking branch 'upstream/dev' into feature/build-diff-viewer

Masusder committed:
Integrated blueprint decompilation with diff viewer

Masusder committed:
Switch provider

Masusder committed:
Merge remote-tracking branch 'upstream/dev' into feature/build-diff-viewer

Masusder committed:
Merge remote-tracking branch 'upstream/dev' into feature/build-diff-viewer

Masusder committed:
Nvm, now searchbar is focusing properly

0belous commented:
Bump this
I would love to see this feature
